use simple ret_ if there is no way of leaking the return statement

o- opened the pull request.

o- commented:
that does not seem to work. I need to investigate. My assumption was that `return` as a top-level statement in a function doesn't do anything funky, but apparently there is something going on.

JanJecmen commented:
I think your idea is fine, it's just a missed case in the verifier:
(btw should `deopt_` also be covered by `isReturn()`?)

```diff
diff --git a/rir/src/ir/CodeVerifier.cpp b/rir/src/ir/CodeVerifier.cpp
index 734c01a..d956555 100644
--- a/rir/src/ir/CodeVerifier.cpp
+++ b/rir/src/ir/CodeVerifier.cpp
@@ -46,7 +46,7 @@ class State {

     void advance() {
         BC bc = BC::advance(&pc);
-        if (bc.bc == Opcode::return_ || bc.bc == Opcode::deopt_)
+        if (bc.isReturn() || bc.bc == Opcode::deopt_)
             ostack = 0;
         else
             ostack -= bc.popCount();
@@ -201,8 +201,7 @@ void CodeVerifier::calculateAndVerifyStack(Code* code) {
             BC cur = BC::decode(pc);
             i.advance();
             max.updateMax(i);
-            if (cur.bc == Opcode::ret_ || cur.bc == Opcode::return_ ||
-                cur.bc == Opcode::deopt_) {
+            if (cur.isReturn() || cur.bc == Opcode::deopt_) {
                 i.checkClear();
                 break;
             } else if (cur.bc == Opcode::br_) {
```

o- commented:
@JanJecmen thanks for looking at it. That is indeed one of the problems. The next problem is, that I do not empty the stack before `ret_`. But there is a problem: in `ir/Compiler.cpp` we do not actually keep track of how high the stack is, from what I can see :(

JanJecmen commented:
Oh right, because until now we didn't care, the `return_` gets the correct stack pointer from the target context...

o- committed:
optimize return

translate return_ to a Pir local return, if the return statement
appears as toplevel statement (thus cannot leak as promise).

o- commented:
let's try to just do this in pir. keeping track of stack height in rir would be a bit of a painful change (though maybe eventually we still want it..)

o- commented:
(I also decided to remove some callback confusion in the code...)

o- commented:
hmm, also actually the verifier should verify that the stack is empty on ret_, so I have to revert that again

JanJecmen commented on rir/src/ir/BC_inc.h:
```diff
@@ -229,7 +229,8 @@ class BC {
     bool isReturn() const {
-        return bc == Opcode::ret_ || bc == Opcode::return_;
+        return bc == Opcode::ret_ || bc == Opcode::return_ ||
+               bc == Opcode::deopt_;
```
Maybe should now be named `isExit` or similar

JanJecmen commented on rir/src/compiler/translations/rir_2_pir/rir_2_pir.h:
```diff
@@ -18,50 +18,27 @@ class Rir2Pir {
+            finalize(res, insert);
+            return true;
+        }
+        return false;
```
Maybe move to `.cpp` to be consistent

JanJecmen commented on rir/src/compiler/translations/rir_2_pir/rir_2_pir.h:
```diff
@@ -75,6 +52,17 @@ class Rir2Pir {
     bool compileBC(const BC& bc, Opcode* pos, rir::Code* srcCode, RirStack&,
                    Builder&, std::unordered_map<Value*, CallFeedback>&,
                    std::unordered_map<Value*, TypeFeedback>&) const;
+    virtual bool inPromise() const { return false; }
```
Wasn't sure why this is private, I like this answer: https://stackoverflow.com/a/2170696
(originally I thought you cannot override private because you don't see it)

JanJecmen commented on rir/src/compiler/translations/rir_2_pir/rir_2_pir.h:
```diff
@@ -18,50 +18,27 @@ class Rir2Pir {
-                               },
-                               []() { return false; });
+        if (auto res = tryTranslate(srcCode, insert)) {
+            finalize(res, insert);
```
Can `finalize` fail? Ie. if the `InsertCast` pass fails, we just abort.. I guess in general I am not sure when an abort is appropriate and when just signalling failure is fine

o- replied:
Finalize should not fail.

tryCompile fails = unsupported feature
finalize fails = bug

for example the compiler should always generate code such that insertcast succeeds. We do insert cast lazily, because sometimes it's hard to do it eager, as you do not know all the usages of a value when you create it

JanJecmen commented on rir/src/compiler/translations/rir_2_pir/rir_2_pir.cpp:
```diff
@@ -696,9 +698,12 @@ void Rir2Pir::translate(rir::Code* srcCode, Builder& insert,
+                    return nullptr;
+                }
+                break;
             default:
```
Is aborting on `deopt_` what we want?

o- replied:
good catch

JanJecmen reviewed (commented):
Looks good, I think removing the callbacks makes it a bit easier to read.
But, I don't see the actual optimization in Compiler.cpp now ;)

o- commented:
>  But, I don't see the actual optimization in Compiler.cpp now ;)

It's a bit hidden, but basically this fallthrough case here:
https://github.com/reactorlabs/rir/pull/223/files#diff-a3b8fcdd57c845ed78e9242b24c2ca2eR705

we will not abort compilation on `return_` if we are outside of the promise, and we will then return tos later in the function. I will make sure to make it more explicit!

JanJecmen commented:
Ah, sneaky
